### ISSUE TITLE
Gradient and Jacobians inside a Module

### ISSUE THREAD

**Issue by keunhong** [google/flax#735]
I'd like to compute gradients and Jacobians inside a module e.g., compute the Jacobian of the outputs of an MLP wr.t. the inputs.

Currently, using the vanilla `jax.jacfwd` does seem to work as intended:

```python
class MLP(nn.Module):
  features: Sequence[int]

  @nn.compact
  def __call__(self, x):
    for feat in self.features[:-1]:
      x = nn.relu(nn.Dense(feat)(x))
    x = nn.Dense(self.features[-1])(x)
    return x


class JacModule(nn.Module):
  features: Sequence[int]

  def setup(self):
    self.mlp = MLP(self.features)

  def __call__(self, x):
    return jax.jacfwd(self.mlp)(x)

model = JacModule([128, 3])
key1, key2, key3 = random.split(random.PRNGKey(0), 3)
x = random.normal(key2, (3,)) # Dummy input
params = model.init(key3, x) # Initialization call
model.apply(params, x)
```

but it would be nice to be able to 'wrap' a module to compute the Jacobian/gradients of the module the same way you can with `nn.vmap` (with the option to return both the output and the Jacobian).

**Comment by gerdm** [comment]
As a followup to @keunhong's question, is the "proper" way to work with both the MLP and the Jacobian of the MLP in Flax to define the same class twice but with different `__call__` method? For example, having

```python
def __call__(self, x):
    return self.mlp(x)
```
for the MLP and,

```python
def __call__(self, x):
    return jax.jacfwd(self.mlp)(x)
```

for the Jacobian of the MLP.

**Comment by jatentaki** [comment]
What do you mean by not working? I just tried this code with `jax==0.2.16, jaxlib==0.1.68` and it seems to work on my side. Has this issue been resolved meanwhile?

**Comment by marcvanzee** [comment]
The code works, but I think @keunhong is asking whether we can also create a lifted transform for computing Jacobians, rather than putting it into `__call__`. @jheek is this something we currently have on our radar?

**Comment by jheek** [comment]
I have been wanting to implement this for quite a while. It's not a big issue it's just that not a lot of people have asked for gradient calculations to be lifted

**Comment by gerdm** [comment]
+1 for the implementation. I have been following this thread for a while, but I wasn't aware that an implementation might be possible.

**Comment by LeonhardFeiner** [comment]
+1 I would be interested in lifted versions of `jacfwd`, `jacref`, `grad`, `value_and_grad` as well.

**Comment by jheek** [comment]
We now have jvp and vjp which are general enough to support all gradient computations (jacfwd, jacref, grad, hessian, etc..)

**Comment by LeonhardFeiner** [comment]
@jheek I don't agree on the generality of jvp and vjp.
In my opinion, it is not possible to retrieve the square of the jacobian and multiply it with another vector, is it?

Here's a dummy calculation (without using flax) for that:
```
def foo(x):
    return (x ** jnp.arange(3 * len(x)).reshape(3, -1)).sum(-1)

x = jnp.arange(-4., 0)
jacobian = jax.jacrev(foo)(x)
jnp.square(jacobian) @ jnp.ones(4)
```
I think neither jvp nor vjp can be used instead of jacrev for this.
This is the case for every non-affine transformation of the jacobian.
